libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -213,19 +261,93 @@ void WaveTrack::Interval::StretchRightTo(double t)
 {
    const auto channelsCount = NChannels();
-
    // It is safe to assume that the track has at least one channel
    assert(channelsCount > 0);
 
-   for (unsigned channel = 0; channel < channelsCount; ++channel)
-      GetClip(channel)->ApplyStretchRatio(
-         [&reportProgress, channel, channelsCount](double progress) {
-            reportProgress((channel + progress) / channelsCount);
-         });
+   const auto originalPlayStartTime = GetPlayStartTime();
+   const auto originalPlayEndTime = GetPlayEndTime();
+   const auto stretchRatio = GetStretchRatio();
+
+   auto success = false;
+   Finally Do { [&] {
+      if (!success)
+      {
+         TrimLeftTo(originalPlayStartTime);
+         TrimRightTo(originalPlayEndTime);
+      }
+   } };
+
+   // Leave 1 second of raw, unstretched audio before and after visible region
+   // to give the algorithm a chance to be in a steady state when reaching the
+   // play boundaries.
+   const auto tmpPlayStartTime =
+      std::max(GetSequenceStartTime(), originalPlayStartTime - stretchRatio);
+   const auto tmpPlayEndTime =
+      std::min(GetSequenceEndTime(), originalPlayEndTime + stretchRatio);
+   TrimLeftTo(tmpPlayStartTime);
+   TrimRightTo(tmpPlayEndTime);
+
+   WideClip wideClip { mpClip, mpClip1 };
+
+   constexpr auto sourceDurationToDiscard = 0.;
+   constexpr auto blockSize = 1024;
+   const auto numChannels = NChannels();
+   ClipTimeAndPitchSource stretcherSource { wideClip, sourceDurationToDiscard,
+                                            PlaybackDirection::forward };
+   TimeAndPitchInterface::Parameters params;
+   params.timeRatio = stretchRatio;
+   StaffPadTimeAndPitch stretcher { wideClip.GetRate(), numChannels, stretcherSource,
+                                    std::move(params) };
+
+   // Post-rendering sample counts, i.e., stretched units
+   const auto totalNumOutSamples =
+      sampleCount { wideClip.GetVisibleSampleCount().as_double() *
+                    stretchRatio };
+
+   sampleCount numOutSamples { 0 };
+   AudioContainer container(blockSize, numChannels);
+
+   const auto dst = std::make_shared<Interval>(
+      group, NChannels(), factory, GetClip(0)->GetRate());
+
+   while (numOutSamples < totalNumOutSamples)
+   {
+      const auto numSamplesToGet =
+         limitSampleBufferSize(blockSize, totalNumOutSamples - numOutSamples);
+      stretcher.GetSamples(container.Get(), numSamplesToGet);
+      constSamplePtr data[2];
+      data[0] = reinterpret_cast<constSamplePtr>(container.Get()[0]);
+      if (NChannels() == 2)
+         data[1] = reinterpret_cast<constSamplePtr>(container.Get()[1]);
+      // TODO I don't think so, but double-checking: the `excsafety{Strong}`
+      // comment of `Sequence::AppendNewBlock` doesn't mean the method can't
+      // throw, right? If it didn't throw, then we could get rid of the Finally
+      // and restore the trimming of `this` after this loop before the call to
+      // `Flush`.
```

Comment:
Answering:  Exception safety was carefully documented for those Sequence operations precisely because exceptions are possble (otherwise it would have been `excsafety{No-Fail}`).  It is a promise that in case of exceptions, the Sequence object is restored to its prior state.

A known problem with wide wave clips (such as we have them, with the WaveTrack::Interval facade) is that strong exception safety for the individual sequences does not imply strong exception safety of the wide clip as a whole when appending or flushing.  Maybe there is success for one channel but failure for the other, and then there is no reversion of the first channel to its previous state.

I did not attempt to solve this unlikely problem for 3.4 and neither should you.  Just maybe it will happen if you record and run out of space on the drive.  There is exception handling in this case to save as much of the recording as possible rather than throw away the entire recording.  This hole in strong exception safety might mean it's one remaining way to end up with mis-aligned clips.